collection_encoders.go

```diff
@@ -0,0 +1,196 @@
+package rdb
+
+import (
+	"math"
+	"time"
+)
+
+type CollectionEncoder interface {
+	WriteZeroLength() error
+
+	WriteFieldStrStr(key string, value string) error
+
+	WriteFieldStrStrWithExpiry(key string, value string, expiry *time.Time) error
+
+	WriteFieldStr(field string) error
+
+	WriteFieldStrFloat64(field string, value float64) error
+
+	Close() error
+}
+
+type baseCollectionEncoder struct {
+	encoder   *Encoder
+	length    int64
+	lengthPos int64
+}
+
+func (s *baseCollectionEncoder) WriteZeroLength() error {
+	startPos, _ := s.encoder.writer.Pos()
+	err := s.encoder.writer.WriteLengthUint64(0)
+	if err != nil {
+		return err
+	}
+	s.length = 0
+	s.lengthPos = startPos
+	return nil
+}
+
+func (s *baseCollectionEncoder) Close() error {
+	finalPos, err := s.encoder.writer.Pos()
+	if err != nil {
+		return err
+	}
+	_, err = s.encoder.writer.SeekPos(s.lengthPos)
+	if err != nil {
+		return err
+	}
+	err = s.encoder.writer.WriteLengthUint64(uint64(s.length))
+	if err != nil {
+		return err
+	}
+	_, err = s.encoder.writer.SeekPos(finalPos)
+	s.encoder.begin.Store(false)
+	return err
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrStr(key string, value string) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStr(field string) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrFloat64(field string, value float64) error {
+	panic("implement me")
+}
+
+func (s *baseCollectionEncoder) WriteFieldStrStrWithExpiry(key string, value string, expiry *time.Time) error {
+	panic("implement me")
+}
+
+type ListEncoder struct {
+	baseCollectionEncoder
+}
+
+func NewListEncoder(e *Encoder) (*ListEncoder, error) {
+	encoder := &ListEncoder{}
+	encoder.encoder = e
+	err := encoder.WriteZeroLength()
+	return encoder, err
+}
+
+func (s *ListEncoder) WriteFieldStr(val string) error {
+	s.length++
+	return s.encoder.writeString(val)
```

Comment:
minor: for consistency with others, where we don't increment the length when the file encoder returns an error, we can refactor this a bit